riscv-page-tables/src/page_table.rs

```diff
@@ -625,6 +625,100 @@ impl<T: PagingMode> PageTableInner<T> {
     }
 }
 
+/// A paging hierarchy for a given addressing type.
+/// This is used for HS and HU paging modes such as Sv48.
+///
+/// TODO: Support non-4k page sizes.
+pub struct FirstStagePageTable<T: PagingMode> {
+    inner: Mutex<PageTableInner<T>>,
+}
+
+impl<T: PagingMode> FirstStagePageTable<T> {
+    /// Creates a new page table root from the provided `root` that must be at least
+    /// `T::root_level().table_pages()` in length and aligned to `T::TOP_LEVEL_ALIGN`.
+    pub fn new(root: SequentialPages<InternalClean>) -> Result<Self> {
+        let inner = PageTableInner::new(root)?;
+        Ok(Self {
+            inner: Mutex::new(inner),
+        })
+    }
+
+    /// Returns the address of the top level page table. The PFN of this address is what should be
+    /// written to the SATP or HGATP CSR to start using the translations provided by this page
+    /// table.
+    pub fn get_root_address(&self) -> SupervisorPageAddr {
+        self.inner.lock().root.base()
+    }
+
+    /// Prepares for mapping `num_pages` pages of size `page_size` starting at `addr` in the mapped
+    /// address space by locking the target PTEs and populating any intermediate page tables using
+    /// `get_pte_page`. Upon success, returns a `PageTableMapper` that is guaranteed to be able to
+    /// map the specified range.
+    pub fn map_range(
+        &self,
+        addr: PageAddr<T::MappedAddressSpace>,
+        page_size: PageSize,
+        num_pages: u64,
+        get_pte_page: &mut dyn FnMut() -> Option<Page<InternalClean>>,
+    ) -> Result<PageTableAddressMapper<T>> {
+        if page_size.is_huge() {
+            return Err(Error::PageSizeNotSupported(page_size));
+        }
+
+        let mut inner = self.inner.lock();
+        for a in addr.iter_from().take(num_pages as usize) {
+            inner.lock_4k_leaf_for_mapping(a, get_pte_page)?;
+        }
+
+        Ok(PageTableAddressMapper::new(self, addr, num_pages))
+    }
+}
+
+/// A range of mapped address space that has been locked for mapping. The PTEs are unlocked when
+/// this struct is dropped. Mapping a page in this range is guaranteed to succeed as long as the
+/// address hasn't already been mapped by this `PageTableAddressMapper`.
+pub struct PageTableAddressMapper<'a, T: PagingMode> {
```

Comment:
bikeshed: Maybe make it more obvious which `Mapper` is used for what by calling them `FirstStageMapper` and `SecondStageMapper` (or similar)